File: ruby/022.rb
# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
  return [] if n == 0
  return ["()"] if n == 1

  parent_results = generate_parenthesis(n - 1)
  results = []
  parent_results.each do |r|
    r.each_char.with_index do |char, i|
      if char == "("
        if i == 0
          results.push("()" + r)
          results.push("(()" + r[1..-1]);
        else
          results.push(r[0..i-1] + "()" + r[i..-1])
          results.push(r[0..i] + "()" + r[i+1..-1])
        end
      end
    end
  end
  results.uniq
end

File: javascript/022.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
  const generateParenthesisSet = function(n) {
    if (n === 1) return {"()": true};
    const set = generateParenthesisSet(n - 1);
    const newSet = {};
    for (var str in set) {
      for (var i = 0; i < str.length; ++i) {
        if (str[i] === "(") {
          newSet[str.substring(0, i + 1) + "()" + str.substring(i + 1)] = true;
          newSet[str.substring(0, i) + "()" + str.substring(i)] = true;
        }
      }
    }

    return newSet;
  }

  return Object.keys(generateParenthesisSet(n));
};